File: src/main/java/com/chidozie/core/normalmap/NormalMappedObjLoader.java
package com.chidozie.core.normalmap;

import com.chidozie.core.renderEngine.Loader;
import net.adventuregame.models.RawModel;
import org.joml.Vector2f;
import org.joml.Vector3f;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class NormalMappedObjLoader {

		public static RawModel loadOBJ(String objFileName, Loader loader) {
		InputStream isr = null;
		InputStream objFile = NormalMappedObjLoader.class.getResourceAsStream("/assets/adventuregame/models/" + objFileName + ".obj");
		if (objFile == null) {
			System.err.println("File not found in class path; don't use any extension");
			return null;
		}
		BufferedReader reader = new BufferedReader(new InputStreamReader(objFile));
		String line;
		List<VertexNM> vertices = new ArrayList<>();
		List<Vector2f> textures = new ArrayList<>();
		List<Vector3f> normals = new ArrayList<>();
		List<Integer> indices = new ArrayList<>();
		try {
			while (true) {
				line = reader.readLine();
				if (line == null) {
					break;
				}
				if (line.startsWith("v ")) {
					String[] currentLine = line.split(" ");
					Vector3f vertex = new Vector3f(
							Float.parseFloat(currentLine[1]),
							Float.parseFloat(currentLine[2]),
							Float.parseFloat(currentLine[3])
					);
					VertexNM newVertex = new VertexNM(vertices.size(), vertex);
					vertices.add(newVertex);
				} else if (line.startsWith("vt ")) {
					String[] currentLine = line.split(" ");
					Vector2f texture = new Vector2f(
							Float.parseFloat(currentLine[1]),
							Float.parseFloat(currentLine[2])
					);
					textures.add(texture);
				} else if (line.startsWith("vn ")) {
					String[] currentLine = line.split(" ");
					Vector3f normal = new Vector3f(
							Float.parseFloat(currentLine[1]),
							Float.parseFloat(currentLine[2]),
							Float.parseFloat(currentLine[3])
					);
					normals.add(normal);
				} else if (line.startsWith("f ")) {
					break;
				}
			}
			while (line != null && line.startsWith("f ")) {
				String[] currentLine = line.split(" ");
				String[] vertex1 = currentLine[1].split("/");
				String[] vertex2 = currentLine[2].split("/");
				String[] vertex3 = currentLine[3].split("/");
				VertexNM v0 = processVertex(vertex1, vertices, indices);
				VertexNM v1 = processVertex(vertex2, vertices, indices);
				VertexNM v2 = processVertex(vertex3, vertices, indices);
				calculateTangents(v0, v1, v2, textures); // NEW
				line = reader.readLine();
			}
			reader.close();
		} catch (IOException e) {
			System.err.println("Error reading the file");
		}
		removeUnusedVertices(vertices);
		float[] verticesArray = new float[vertices.size() * 3];
		float[] texturesArray = new float[vertices.size() * 2];
		float[] normalsArray = new float[vertices.size() * 3];
		float[] tangentsArray = new float[vertices.size() * 3];
		float furthest = convertDataToArrays(vertices, textures, normals, verticesArray,
				texturesArray, normalsArray, tangentsArray);
		int[] indicesArray = convertIndicesListToArray(indices);

		return loader.loadToVAO(verticesArray, texturesArray, normalsArray, tangentsArray, indicesArray);
	}

	//NEW 
	private static void calculateTangents(VertexNM v0, VertexNM v1, VertexNM v2, List<Vector2f> textures) {
		Vector3f deltaPos1 = new Vector3f(v1.getPosition()).sub(v0.getPosition());
		Vector3f deltaPos2 = new Vector3f(v2.getPosition()).sub(v0.getPosition());
		Vector2f uv0 = textures.get(v0.getTextureIndex());
		Vector2f uv1 = textures.get(v1.getTextureIndex());
		Vector2f uv2 = textures.get(v2.getTextureIndex());
		Vector2f deltaUv1 = new Vector2f(uv1).sub(uv0);
		Vector2f deltaUv2 = new Vector2f(uv2).sub(uv0);

		float r = 1.0f / (deltaUv1.x * deltaUv2.y - deltaUv1.y * deltaUv2.x);
		deltaPos1.mul(deltaUv2.y);
		deltaPos2.mul(deltaUv1.y);
		Vector3f tangent = deltaPos1.sub(deltaPos2).mul(r);
		v0.addTangent(tangent);
		v1.addTangent(tangent);
		v2.addTangent(tangent);
	}

	private static VertexNM processVertex(String[] vertex, List<VertexNM> vertices,
			List<Integer> indices) {
		int index = Integer.parseInt(vertex[0]) - 1;
		VertexNM currentVertex = vertices.get(index);
		int textureIndex = Integer.parseInt(vertex[1]) - 1;
		int normalIndex = Integer.parseInt(vertex[2]) - 1;
		if (!currentVertex.isSet()) {
			currentVertex.setTextureIndex(textureIndex);
			currentVertex.setNormalIndex(normalIndex);
			indices.add(index);
			return currentVertex;
		} else {
			return dealWithAlreadyProcessedVertex(currentVertex, textureIndex, normalIndex, indices,
					vertices);
		}
	}

	private static int[] convertIndicesListToArray(List<Integer> indices) {
		int[] indicesArray = new int[indices.size()];
		for (int i = 0; i < indicesArray.length; i++) {
			indicesArray[i] = indices.get(i);
		}
		return indicesArray;
	}

	private static float convertDataToArrays(List<VertexNM> vertices, List<Vector2f> textures,
			List<Vector3f> normals, float[] verticesArray, float[] texturesArray,
			float[] normalsArray, float[] tangentsArray) {
		float furthestPoint = 0;
		for (int i = 0; i < vertices.size(); i++) {
			VertexNM currentVertex = vertices.get(i);
			if (currentVertex.getLength() > furthestPoint) {
				furthestPoint = currentVertex.getLength();
			}
			Vector3f position = currentVertex.getPosition();
			Vector2f textureCoord = textures.get(currentVertex.getTextureIndex());
			Vector3f normalVector = normals.get(currentVertex.getNormalIndex());
			Vector3f tangent = currentVertex.getAverageTangent();
			verticesArray[i * 3] = position.x;
			verticesArray[i * 3 + 1] = position.y;
			verticesArray[i * 3 + 2] = position.z;
			texturesArray[i * 2] = textureCoord.x;
			texturesArray[i * 2 + 1] = 1 - textureCoord.y;
			normalsArray[i * 3] = normalVector.x;
			normalsArray[i * 3 + 1] = normalVector.y;
			normalsArray[i * 3 + 2] = normalVector.z;
			tangentsArray[i * 3] = tangent.x;
			tangentsArray[i * 3 + 1] = tangent.y;
			tangentsArray[i * 3 + 2] = tangent.z;

		}
		return furthestPoint;
	}

	private static VertexNM dealWithAlreadyProcessedVertex(VertexNM previousVertex, int newTextureIndex,
			int newNormalIndex, List<Integer> indices, List<VertexNM> vertices) {
		if (previousVertex.hasSameTextureAndNormal(newTextureIndex, newNormalIndex)) {
			indices.add(previousVertex.getIndex());
			return previousVertex;
		} else {
			VertexNM anotherVertex = previousVertex.getDuplicateVertex();
			if (anotherVertex != null) {
				return dealWithAlreadyProcessedVertex(anotherVertex, newTextureIndex,
						newNormalIndex, indices, vertices);
			} else {
				VertexNM duplicateVertex = previousVertex.duplicate(vertices.size());//NEW
				duplicateVertex.setTextureIndex(newTextureIndex);
				duplicateVertex.setNormalIndex(newNormalIndex);
				previousVertex.setDuplicateVertex(duplicateVertex);
				vertices.add(duplicateVertex);
				indices.add(duplicateVertex.getIndex());
				return duplicateVertex;
			}
		}
	}

	private static void removeUnusedVertices(List<VertexNM> vertices) {
		for (VertexNM vertex : vertices) {
			vertex.averageTangents();
			if (!vertex.isSet()) {
				vertex.setTextureIndex(0);
				vertex.setNormalIndex(0);
			}
		}
	}

}

File: src/main/kotlin/com/chidozie/core/normalmap/NormalMappedObjLoader.kt
package com.chidozie.core.normalmap

import com.chidozie.core.renderEngine.Loader
import net.adventuregame.models.RawModel
import org.joml.Vector2f
import org.joml.Vector3f
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader

object NormalMappedObjLoader {
    fun loadOBJ(objFileName: String?, loader: Loader): RawModel? {
        val isr: InputStream? = null
        val objFile =
            NormalMappedObjLoader::class.java.getResourceAsStream("/assets/adventuregame/models/" + objFileName + ".obj")
        if (objFile == null) {
            System.err.println("File not found in class path; don't use any extension")
            return null
        }
        val reader = BufferedReader(InputStreamReader(objFile))
        var line: String?
        val vertices: MutableList<VertexNM> = ArrayList<VertexNM>()
        val textures: MutableList<Vector2f> = ArrayList<Vector2f>()
        val normals: MutableList<Vector3f> = ArrayList<Vector3f>()
        val indices: MutableList<Int?> = ArrayList<Int?>()
        try {
            while (true) {
                line = reader.readLine()
                if (line == null) {
                    break
                }
                if (line.startsWith("v ")) {
                    val currentLine: Array<String?> =
                        line.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    val vertex = Vector3f(
                        currentLine[1]!!.toFloat(),
                        currentLine[2]!!.toFloat(),
                        currentLine[3]!!.toFloat()
                    )
                    val newVertex = VertexNM(vertices.size, vertex)
                    vertices.add(newVertex)
                } else if (line.startsWith("vt ")) {
                    val currentLine: Array<String?> =
                        line.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    val texture = Vector2f(
                        currentLine[1]!!.toFloat(),
                        currentLine[2]!!.toFloat()
                    )
                    textures.add(texture)
                } else if (line.startsWith("vn ")) {
                    val currentLine: Array<String?> =
                        line.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    val normal = Vector3f(
                        currentLine[1]!!.toFloat(),
                        currentLine[2]!!.toFloat(),
                        currentLine[3]!!.toFloat()
                    )
                    normals.add(normal)
                } else if (line.startsWith("f ")) {
                    break
                }
            }
            while (line != null && line.startsWith("f ")) {
                val currentLine: Array<String?> =
                    line.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val vertex1: Array<String?> =
                    currentLine[1]!!.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val vertex2: Array<String?> =
                    currentLine[2]!!.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val vertex3: Array<String?> =
                    currentLine[3]!!.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val v0 = processVertex(vertex1, vertices, indices)
                val v1 = processVertex(vertex2, vertices, indices)
                val v2 = processVertex(vertex3, vertices, indices)
                calculateTangents(v0, v1, v2, textures) // NEW
                line = reader.readLine()
            }
            reader.close()
        } catch (e: IOException) {
            System.err.println("Error reading the file")
        }
        removeUnusedVertices(vertices)
        val verticesArray = FloatArray(vertices.size * 3)
        val texturesArray = FloatArray(vertices.size * 2)
        val normalsArray = FloatArray(vertices.size * 3)
        val tangentsArray = FloatArray(vertices.size * 3)
        val furthest = convertDataToArrays(
            vertices, textures, normals, verticesArray,
            texturesArray, normalsArray, tangentsArray
        )
        val indicesArray = convertIndicesListToArray(indices)

        return loader.loadToVAO(verticesArray, texturesArray, normalsArray, tangentsArray, indicesArray)
    }

    //NEW 
    private fun calculateTangents(v0: VertexNM, v1: VertexNM, v2: VertexNM, textures: MutableList<Vector2f>) {
        val deltaPos1 = Vector3f(v1.position).sub(v0.position)
        val deltaPos2 = Vector3f(v2.position).sub(v0.position)
        val uv0: Vector2f? = textures[v0.textureIndex]
        val uv1 = textures[v1.textureIndex]
        val uv2 = textures[v2.textureIndex]
        val deltaUv1 = Vector2f(uv1).sub(uv0)
        val deltaUv2 = Vector2f(uv2).sub(uv0)

        val r = 1.0f / (deltaUv1.x * deltaUv2.y - deltaUv1.y * deltaUv2.x)
        deltaPos1.mul(deltaUv2.y)
        deltaPos2.mul(deltaUv1.y)
        val tangent = deltaPos1.sub(deltaPos2).mul(r)
        v0.addTangent(tangent)
        v1.addTangent(tangent)
        v2.addTangent(tangent)
    }

    private fun processVertex(
        vertex: Array<String?>, vertices: MutableList<VertexNM>,
        indices: MutableList<Int?>
    ): VertexNM {
        val index = vertex[0]!!.toInt() - 1
        val currentVertex = vertices.get(index)
        val textureIndex = vertex[1]!!.toInt() - 1
        val normalIndex = vertex[2]!!.toInt() - 1
        if (!currentVertex.isSet) {
            currentVertex.textureIndex = textureIndex
            currentVertex.normalIndex = normalIndex
            indices.add(index)
            return currentVertex
        } else {
            return dealWithAlreadyProcessedVertex(
                currentVertex, textureIndex, normalIndex, indices,
                vertices
            )
        }
    }

    private fun convertIndicesListToArray(indices: MutableList<Int?>): IntArray {
        val indicesArray = IntArray(indices.size)
        for (i in indicesArray.indices) {
            indicesArray[i] = indices.get(i)!!
        }
        return indicesArray
    }

    private fun convertDataToArrays(
        vertices: MutableList<VertexNM>, textures: MutableList<Vector2f>,
        normals: MutableList<Vector3f>, verticesArray: FloatArray, texturesArray: FloatArray,
        normalsArray: FloatArray, tangentsArray: FloatArray
    ): Float {
        var furthestPoint = 0f
        for (i in vertices.indices) {
            val currentVertex = vertices[i]
            if (currentVertex.length > furthestPoint) {
                furthestPoint = currentVertex.length
            }
            val position = currentVertex.position
            val textureCoord = textures[currentVertex.textureIndex]
            val normalVector = normals[currentVertex.normalIndex]
            val tangent = currentVertex.averageTangent
            verticesArray[i * 3] = position.x
            verticesArray[i * 3 + 1] = position.y
            verticesArray[i * 3 + 2] = position.z
            texturesArray[i * 2] = textureCoord.x
            texturesArray[i * 2 + 1] = 1 - textureCoord.y
            normalsArray[i * 3] = normalVector.x
            normalsArray[i * 3 + 1] = normalVector.y
            normalsArray[i * 3 + 2] = normalVector.z
            tangentsArray[i * 3] = tangent.x
            tangentsArray[i * 3 + 1] = tangent.y
            tangentsArray[i * 3 + 2] = tangent.z
        }
        return furthestPoint
    }

    private fun dealWithAlreadyProcessedVertex(
        previousVertex: VertexNM, newTextureIndex: Int,
        newNormalIndex: Int, indices: MutableList<Int?>, vertices: MutableList<VertexNM>
    ): VertexNM {
        if (previousVertex.hasSameTextureAndNormal(newTextureIndex, newNormalIndex)) {
            indices.add(previousVertex.index)
            return previousVertex
        } else {
            val anotherVertex = previousVertex.duplicateVertex
            if (anotherVertex != null) {
                return dealWithAlreadyProcessedVertex(
                    anotherVertex, newTextureIndex,
                    newNormalIndex, indices, vertices
                )
            } else {
                val duplicateVertex = previousVertex.duplicate(vertices.size) //NEW
                duplicateVertex.textureIndex = newTextureIndex
                duplicateVertex.normalIndex = newNormalIndex
                previousVertex.duplicateVertex = duplicateVertex
                vertices.add(duplicateVertex)
                indices.add(duplicateVertex.index)
                return duplicateVertex
            }
        }
    }

    private fun removeUnusedVertices(vertices: MutableList<VertexNM>) {
        for (vertex in vertices) {
            vertex.averageTangents()
            if (!vertex.isSet) {
                vertex.textureIndex = 0
                vertex.normalIndex = 0
            }
        }
    }
}